internal/runners/packages/rationalize.go

```diff
@@ -0,0 +1,68 @@
+package packages
+
+import (
+	"errors"
+
+	"github.com/ActiveState/cli/internal/errs"
+	"github.com/ActiveState/cli/internal/locale"
+	"github.com/ActiveState/cli/internal/runbits/rationalize"
+	bpModel "github.com/ActiveState/cli/pkg/platform/api/buildplanner/model"
+	"github.com/ActiveState/cli/pkg/platform/authentication"
+	"github.com/ActiveState/cli/pkg/platform/runtime/buildexpression"
+)
+
+func rationalizeError(auth *authentication.Auth, err *error) {
+	var commitError *bpModel.CommitError
+	var requirementNotFoundErr *buildexpression.RequirementNotFoundError
+
+	switch {
+	case err == nil:
+		return
+
+	// No activestate.yaml.
+	case errors.Is(*err, rationalize.ErrNoProject):
+		*err = errs.WrapUserFacing(*err,
+			locale.T("err_no_project"),
+			errs.SetInput(),
+		)
+
+	// Error staging a commit during install.
+	case errors.As(*err, &commitError):
+		switch commitError.Type {
+		case bpModel.NotFoundErrorType:
+			*err = errs.WrapUserFacing(*err,
+				locale.Tl("err_packages_not_found", "Could not make runtime changes because your project was not found."),
+				errs.SetInput(),
+				errs.SetTips(locale.T("tip_private_project_auth")),
+			)
+		case bpModel.ForbiddenErrorType:
+			*err = errs.WrapUserFacing(*err,
+				locale.Tl("err_packages_forbidden", "Could not make runtime changes because you do not have permission to do so."),
+				errs.SetInput(),
+				errs.SetTips(locale.T("tip_private_project_auth")),
+			)
+		case bpModel.HeadOnBranchMovedErrorType:
+			*err = errs.WrapUserFacing(*err,
+				locale.T("err_buildplanner_head_on_branch_moved"),
+				errs.SetInput(),
+			)
+		case bpModel.NoChangeSinceLastCommitErrorType:
+			*err = errs.WrapUserFacing(*err,
+				locale.Tl("err_packages_exists", "That package is already installed."),
```

Comment:
I think this message will have to be more generic as it could also arise when running `state uninstall`.

Reply:
May I ask how? If the user specifies a requirement that is not already installed, they'll get a `requirementNotFoundErr`, which is handled below.

Reply:
Ahh I see, I missed that part. Thanks for pointing that out.